File: frontend/src/features/matrix-editor/matrixSchedulePlanning.ts
export type MatrixSchedulePlan = {
  postTestBufferDays: string;
  sampleReceivedDate: string;
  plannedTestStartDate: string;
  plannedTestCompleteDate: string;
  estimatedCompletionDate: string;
};

export type MatrixScheduleRowInput = {
  id: string;
  isSampleRow: boolean;
  dayExpression: string;
  groups: Record<string, string>;
};

export type MatrixScheduleGroupInput = {
  id: string;
  name: string;
  isSelected: boolean;
};

export type MatrixScheduleCalculation = {
  groupDays: Record<string, number>;
  criticalGroupId: string | null;
  criticalGroupDays: number;
  totalCycleDays: number;
  rowErrors: Record<string, string>;
  bufferErrors: Partial<Record<keyof Pick<MatrixSchedulePlan, "postTestBufferDays">, string>>;
  invalidDateFields: Partial<Record<keyof Pick<
    MatrixSchedulePlan,
    "sampleReceivedDate" | "plannedTestStartDate" | "plannedTestCompleteDate" | "estimatedCompletionDate"
  >, true>>;
  dateError: string | null;
  isValid: boolean;
};

type DateField = keyof Pick<
  MatrixSchedulePlan,
  "sampleReceivedDate" | "plannedTestStartDate" | "plannedTestCompleteDate" | "estimatedCompletionDate"
>;

type DateValidationResult = {
  message: string | null;
  invalidFields: Partial<Record<DateField, true>>;
};

const DATE_FIELDS: DateField[] = [
  "sampleReceivedDate",
  "plannedTestStartDate",
  "plannedTestCompleteDate",
  "estimatedCompletionDate",
];

const DAY_PATTERN = /^\d+(?:\.\d+)?x?$/i;
const DECIMAL_PATTERN = /^\d+(?:\.\d+)?$/;

export function emptySchedulePlan(): MatrixSchedulePlan {
  return {
    postTestBufferDays: "",
    sampleReceivedDate: "",
    plannedTestStartDate: "",
    plannedTestCompleteDate: "",
    estimatedCompletionDate: "",
  };
}

export function calculateMatrixSchedule(
  rows: MatrixScheduleRowInput[],
  groups: MatrixScheduleGroupInput[],
  plan: MatrixSchedulePlan
): MatrixScheduleCalculation {
  const selectedGroups = groups.filter((group) => group.isSelected);
  const groupDays: Record<string, number> = {};
  selectedGroups.forEach((group) => {
    groupDays[group.id] = 0;
  });

  const rowErrors: Record<string, string> = {};
  const activeRowIds = new Set<string>();
  rows.forEach((row) => {
    if (row.isSampleRow) {
      return;
    }
    if (
      selectedGroups.some((group) => countStepTokens(row.groups[group.id] ?? "") > 0)
    ) {
      activeRowIds.add(row.id);
    }
  });

  rows.forEach((row) => {
    if (row.isSampleRow || !activeRowIds.has(row.id)) {
      return;
    }
    const parsed = parseDayExpression(row.dayExpression);
    if (parsed.error) {
      rowErrors[row.id] = parsed.error;
      return;
    }
    if (parsed.value == null) {
      return;
    }
    const dayValue = parsed.value;
    selectedGroups.forEach((group) => {
      const tokenCount = countStepTokens(row.groups[group.id] ?? "");
      if (tokenCount === 0) {
        return;
      }
      groupDays[group.id] += parsed.multiplier ? dayValue * tokenCount : dayValue;
    });
  });

  const bufferErrors: MatrixScheduleCalculation["bufferErrors"] = {};
  const postBuffer = parseBufferDays(plan.postTestBufferDays);
  if (postBuffer.error) {
    bufferErrors.postTestBufferDays = postBuffer.error;
  }

  const critical = selectedGroups.reduce(
    (current, group) =>
      groupDays[group.id] > current.days ? { id: group.id, days: groupDays[group.id] } : current,
    { id: null as string | null, days: 0 }
  );
  const totalCycleDays = critical.days + (postBuffer.value ?? 0);
  const dateValidation =
    Object.keys(bufferErrors).length > 0
      ? { message: null, invalidFields: {} }
      : validateDatePlan(plan, critical.days, postBuffer.value ?? 0);

  return {
    groupDays,
    criticalGroupId: critical.id,
    criticalGroupDays: critical.days,
    totalCycleDays,
    rowErrors,
    bufferErrors,
    invalidDateFields: dateValidation.invalidFields,
    dateError: dateValidation.message,
    isValid:
      Object.keys(rowErrors).length === 0 &&
      Object.keys(bufferErrors).length === 0 &&
      dateValidation.message == null,
  };
}

export function formatPlanningDays(value: number): string {
  if (!Number.isFinite(value)) {
    return "0";
  }
  return Number.isInteger(value) ? String(value) : value.toFixed(2).replace(/0+$/, "").replace(/\.$/, "");
}

function parseDayExpression(value: string): { value: number | null; multiplier: boolean; error: string | null } {
  const text = value.trim();
  if (!text) {
    return { value: null, multiplier: false, error: null };
  }
  if (!DAY_PATTERN.test(text)) {
    return { value: null, multiplier: false, error: "Day must be a non-negative decimal or decimal multiplier." };
  }
  const multiplier = text.toLowerCase().endsWith("x");
  const numericText = multiplier ? text.slice(0, -1) : text;
  const parsed = Number.parseFloat(numericText);
  return { value: parsed, multiplier, error: null };
}

function parseBufferDays(value: string): { value: number | null; error: string | null } {
  const text = value.trim();
  if (!text) {
    return { value: 0, error: null };
  }
  if (!DECIMAL_PATTERN.test(text)) {
    return { value: null, error: "Buffer days must be a non-negative decimal." };
  }
  return { value: Number.parseFloat(text), error: null };
}

function countStepTokens(value: string): number {
  return value.trim().split(/[\s,，、\u040e\u045e]+/).filter((token) => token.trim().length > 0).length;
}

function validateDatePlan(
  plan: MatrixSchedulePlan,
  criticalGroupDays: number,
  postBufferDays: number,
): DateValidationResult {
  const noError: DateValidationResult = { message: null, invalidFields: {} };
  const values = [
    plan.sampleReceivedDate,
    plan.plannedTestStartDate,
    plan.plannedTestCompleteDate,
    plan.estimatedCompletionDate,
  ].map((value) => value.trim());
  if (!values.some(Boolean)) {
    return noError;
  }
  if (!values.every(Boolean)) {
    return {
      message: "All planned date fields are required when any planned date is filled.",
      invalidFields: buildInvalidDateFields(values.map((value) => !value)),
    };
  }
  const [received, start, complete, estimated] = values.map(parseDateValue);
  if (received == null || start == null || complete == null || estimated == null) {
    return {
      message: "Planned dates must use YYYY-MM-DD format.",
      invalidFields: buildInvalidDateFields([
        received == null,
        start == null,
        complete == null,
        estimated == null,
      ]),
    };
  }
  if (start < received) {
    return {
      message: "Planned start is earlier than sample received date.",
      invalidFields: { plannedTestStartDate: true },
    };
  }
  if (complete < addCalendarDays(start, Math.ceil(criticalGroupDays))) {
    return {
      message: "Test complete is earlier than planned start plus critical group days.",
      invalidFields: { plannedTestCompleteDate: true },
    };
  }
  if (estimated < addCalendarDays(complete, Math.ceil(postBufferDays))) {
    return {
      message: "Estimated completion is earlier than test complete plus post-test buffer.",
      invalidFields: { estimatedCompletionDate: true },
    };
  }
  return noError;
}

function parseDateValue(value: string): number | null {
  const isoMatch = value.match(/^(\d{4})-(\d{2})-(\d{2})$/);
  if (isoMatch != null) {
    const [, year, month, day] = isoMatch;
    const date = Date.UTC(Number(year), Number(month) - 1, Number(day));
    if (!Number.isFinite(date)) {
      return null;
    }
    const parsed = new Date(date);
    const parsedYear = String(parsed.getUTCFullYear()).padStart(4, "0");
    const parsedMonth = String(parsed.getUTCMonth() + 1).padStart(2, "0");
    const parsedDay = String(parsed.getUTCDate()).padStart(2, "0");
    return `${parsedYear}-${parsedMonth}-${parsedDay}` === value ? date : null;
  }
  return null;
}

function addCalendarDays(dateValue: number, days: number): number {
  return dateValue + days * 24 * 60 * 60 * 1000;
}

function buildInvalidDateFields(flags: boolean[]): Partial<Record<DateField, true>> {
  return DATE_FIELDS.reduce<Partial<Record<DateField, true>>>((invalidFields, field, index) => {
    if (flags[index]) {
      invalidFields[field] = true;
    }
    return invalidFields;
  }, {});
}

File: backend/application/matrix_schedule_planning.py
"""Deterministic helpers for Matrix planning day and schedule validation."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, timedelta
from decimal import Decimal, InvalidOperation, ROUND_CEILING
import re
from typing import Iterable, Mapping


class MatrixScheduleValidationError(ValueError):
    """Raised when Matrix planning values fail validation."""


@dataclass(frozen=True, slots=True)
class ParsedDayExpression:
    """Parsed row-level planned day expression."""

    value: Decimal
    multiplier: bool


@dataclass(frozen=True, slots=True)
class MatrixScheduleFields:
    """Root-level Matrix planning schedule fields."""

    pre_test_buffer_days: str | None = None
    post_test_buffer_days: str | None = None
    sample_received_date: str | None = None
    planned_test_start_date: str | None = None
    planned_test_complete_date: str | None = None
    estimated_completion_date: str | None = None


@dataclass(frozen=True, slots=True)
class MatrixScheduleValidationResult:
    """Validated schedule calculation result."""

    critical_group_id: str | None
    critical_group_test_days: Decimal
    total_cycle_days: Decimal


def parse_day_expression(value: str | None) -> ParsedDayExpression | None:
    """Parse an editable row Day expression."""
    text = (value or "").strip()
    if not text:
        return None
    multiplier = text.lower().endswith("x")
    number_text = text[:-1] if multiplier else text
    parsed = _parse_non_negative_decimal(number_text, value_name="Day")
    return ParsedDayExpression(value=parsed, multiplier=multiplier)


def parse_buffer_days(value: str | None, *, value_name: str) -> Decimal:
    """Parse a root-level non-negative buffer day value; blank means zero."""
    text = (value or "").strip()
    if not text:
        return Decimal("0")
    if text.lower().endswith("x"):
        raise MatrixScheduleValidationError(f"{value_name} must be a non-negative decimal.")
    return _parse_non_negative_decimal(text, value_name=value_name)


def count_step_tokens(cell_value: str | None) -> int:
    """Count step-like tokens using the Matrix Editor comma/space token convention."""
    text = (cell_value or "").strip()
    if not text:
        return 0
    return len([token for token in re.split(r"[\s,，、\u040e\u045e]+", text) if token.strip()])


def calculate_group_test_days(
    *,
    rows: Iterable[Mapping[str, object]],
    cells: Iterable[Mapping[str, object]],
    selected_group_ids: Iterable[str],
) -> dict[str, Decimal]:
    """Calculate planned test days by selected group id."""
    selected = [group_id for group_id in selected_group_ids if group_id]
    totals = {group_id: Decimal("0") for group_id in selected}
    selected_set = set(selected)
    active_row_ids: set[str] = set()
    cell_payloads: list[Mapping[str, object]] = []
    for cell in cells:
        cell_payloads.append(cell)
        row_id = str(cell.get("row_id") or cell.get("draft_row_id") or "")
        group_id = str(cell.get("group_id") or cell.get("draft_group_id") or "")
        if group_id not in selected_set:
            continue
        if count_step_tokens(_optional_string(cell.get("cell_value"))) > 0:
            active_row_ids.add(row_id)
    row_day_by_id: dict[str, ParsedDayExpression] = {}
    for row in rows:
        if bool(row.get("is_sample_row")) or bool(row.get("is_information_row")):
            continue
        row_id = str(row.get("row_id") or row.get("draft_row_id") or "")
        if not row_id or row_id not in active_row_ids:
            continue
        parsed = parse_day_expression(_optional_string(row.get("day_expression")))
        if parsed is not None:
            row_day_by_id[row_id] = parsed
    for cell in cell_payloads:
        row_id = str(cell.get("row_id") or cell.get("draft_row_id") or "")
        group_id = str(cell.get("group_id") or cell.get("draft_group_id") or "")
        if group_id not in selected_set:
            continue
        parsed = row_day_by_id.get(row_id)
        if parsed is None:
            continue
        token_count = count_step_tokens(_optional_string(cell.get("cell_value")))
        if token_count == 0:
            continue
        contribution = parsed.value * Decimal(token_count) if parsed.multiplier else parsed.value
        totals[group_id] += contribution
    return totals


def validate_planned_schedule(
    *,
    fields: MatrixScheduleFields,
    group_test_days: Mapping[str, Decimal],
) -> MatrixScheduleValidationResult:
    """Validate root schedule fields against calculated group days."""
    post_days = parse_buffer_days(fields.post_test_buffer_days, value_name="Post-test buffer days")
    critical_group_id, critical_days = _critical_group(group_test_days)
    total_cycle_days = critical_days + post_days
    dates = _parse_planned_dates(fields)
    if dates is None:
        return MatrixScheduleValidationResult(
            critical_group_id=critical_group_id,
            critical_group_test_days=critical_days,
            total_cycle_days=total_cycle_days,
        )
    received, start, complete, estimated = dates
    _require_date_at_least(
        actual=start,
        minimum=received,
        message="planned_test_start_date is earlier than sample_received_date.",
    )
    _require_date_at_least(
        actual=complete,
        minimum=start + timedelta(days=_ceil_days(critical_days)),
        message="planned_test_complete_date is earlier than planned_test_start_date plus critical group test days.",
    )
    _require_date_at_least(
        actual=estimated,
        minimum=complete + timedelta(days=_ceil_days(post_days)),
        message="estimated_completion_date is earlier than planned_test_complete_date plus post-test buffer days.",
    )
    return MatrixScheduleValidationResult(
        critical_group_id=critical_group_id,
        critical_group_test_days=critical_days,
        total_cycle_days=total_cycle_days,
    )


def _parse_non_negative_decimal(value: str, *, value_name: str) -> Decimal:
    text = value.strip()
    if not re.fullmatch(r"\d+(?:\.\d+)?", text):
        raise MatrixScheduleValidationError(f"{value_name} must be a non-negative decimal.")
    try:
        parsed = Decimal(text)
    except InvalidOperation as exc:
        raise MatrixScheduleValidationError(f"{value_name} must be a non-negative decimal.") from exc
    if parsed < 0:
        raise MatrixScheduleValidationError(f"{value_name} must be non-negative.")
    return parsed


def _parse_planned_dates(
    fields: MatrixScheduleFields,
) -> tuple[date, date, date, date] | None:
    values = [
        fields.sample_received_date,
        fields.planned_test_start_date,
        fields.planned_test_complete_date,
        fields.estimated_completion_date,
    ]
    filled = [(value or "").strip() for value in values]
    if not any(filled):
        return None
    if not all(filled):
        raise MatrixScheduleValidationError(
            "All planned date fields are required when any planned date is filled."
        )
    try:
        return tuple(date.fromisoformat(value) for value in filled)  # type: ignore[return-value]
    except ValueError as exc:
        raise MatrixScheduleValidationError("Planned dates must use YYYY-MM-DD format.") from exc


def _critical_group(group_test_days: Mapping[str, Decimal]) -> tuple[str | None, Decimal]:
    if not group_test_days:
        return None, Decimal("0")
    group_id, days = max(group_test_days.items(), key=lambda item: (item[1], item[0]))
    return group_id, days


def _ceil_days(value: Decimal) -> int:
    if value <= 0:
        return 0
    return int(value.to_integral_value(rounding=ROUND_CEILING))


def _require_date_at_least(*, actual: date, minimum: date, message: str) -> None:
    if actual < minimum:
        raise MatrixScheduleValidationError(message)


def _optional_string(value: object) -> str | None:
    return value if isinstance(value, str) else None
